py/desiutil/bitmask.py

```diff
@@ -0,0 +1,165 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+# -*- coding: utf-8 -*-
+"""
+================
+desiutil.bitmask
+================
+
+Mask bits for the spectro pipeline.
+
+Individual packages will define their own mask bits and use this as a utility
+access wrapper.  Typical users will not need to construct BitMask objects on
+their own.
+
+Stephen Bailey, Lawrence Berkeley National Lab
+Fall 2015
+
+Example::
+
+    #  Creating a BitMask
+    from desiutil.bitmask import BitMask
+
+    import yaml
+    _bitdefs = yaml.load('''
+    ccdmask:
+        - [BAD,       0, "Pre-determined bad pixel (any reason)"]
+        - [HOT,       1, "Hot pixel"]
+        - [DEAD,      2, "Dead pixel"]
+        - [SATURATED, 3, "Saturated pixel from object"]
+        - [COSMIC,    4, "Cosmic ray"]
+    ''')
+    ccdmask = BitMask('ccdmask', _bitdefs)
+
+    #  Accessing the mask
+    ccdmask.COSMIC | ccdmask.SATURATED  #  2**4 + 2**3
+    ccdmask.mask('COSMIC')     #  2**4, same as ccdmask.COSMIC
+    ccdmask.mask(4)            #  2**4, same as ccdmask.COSMIC
+    ccdmask.COSMIC             #  2**4, same as ccdmask.mask('COSMIC')
+    ccdmask.bitnum('COSMIC')   #  4
+    ccdmask.bitname(4)         #  'COSMIC'
+    ccdmask.names()            #  ['BAD', 'HOT', 'DEAD', 'SATURATED', 'COSMIC']
+    ccdmask.names(3)           #  ['BAD', 'HOT']
+    ccdmask.comment(0)         #  "Pre-determined bad pixel (any reason)"
+    ccdmask.comment('COSMIC')  #  "Cosmic ray"
+"""
+
+
+class _MaskBit(int):
+    """
+    A single mask bit.  Subclasses int to act like an int, but allows the
+    ability to extend with blat.name, blat.comment, blat.mask, blat.bitnum.
+    """
+    def __new__(cls, name, bitnum, comment, extra=dict()):
+        self = super(_MaskBit, cls).__new__(cls, 2**bitnum)
+        self.name = name
+        self.bitnum = bitnum
+        self.mask = 2**bitnum
+        self.comment = comment
+        for key, value in extra.items():
+            assert key not in (
```

Comment:
There are  serious problems with this assertion.  First, `int.bitlength` is not actually an attribute of `int`.  It is `int.bit_length`.  Second, this assumes that the set of attributes of `int` will never change.  Finally, it ignores any of the special attributes of `int`.  With this code, one could in principle override `int.__add__`, which I don't think is what was intended.

So, is there any need to handle extra keywords *at all*?  If so, they should be passed (via `**kwargs`) to the base class for handling, then you can use `hasattr()` to test to see whether the attribute already exists, and *then* raise an error if necessary.